File: app/src/main/java/com/veyndan/paper/reddit/FilterFragment.java
package com.veyndan.paper.reddit;

import android.content.Intent;
import android.graphics.Color;
import android.graphics.PorterDuff;
import android.graphics.drawable.Drawable;
import android.os.Bundle;
import android.support.design.widget.TabLayout;
import android.support.v4.app.DialogFragment;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.FragmentStatePagerAdapter;
import android.support.v4.content.ContextCompat;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.jakewharton.rxbinding2.support.design.widget.RxTabLayout;
import com.jakewharton.rxbinding2.view.RxView;
import com.veyndan.paper.reddit.databinding.FragmentFilterBinding;

import timber.log.Timber;

public class FilterFragment extends DialogFragment {

    @SuppressWarnings("RedundantNoArgConstructor")
    public FilterFragment() {
        // Required empty public constructor
    }

    public static FilterFragment newInstance() {
        return new FilterFragment();
    }

    @Override
    public View onCreateView(final LayoutInflater inflater, final ViewGroup container,
                             final Bundle savedInstanceState) {
        final FragmentFilterBinding binding = FragmentFilterBinding.inflate(inflater, container, false);

        final Fragment[] fragments = {
                TimePeriodFilterFragment.newInstance(),
                SubredditFilterFragment.newInstance(),
                UserFilterFragment.newInstance()
        };

        RxView.clicks(binding.filterDone)
                .subscribe(aVoid -> {
                    final Intent intent = new Intent(getContext(), MainActivity.class);
                    for (final Fragment fragment : fragments) {
                        intent.putExtras(((Filter) fragment).requestFilter());
                    }
                    startActivity(intent);

                    dismiss();
                }, Timber::e);

        final FragmentManager fragmentManager = getChildFragmentManager();
        binding.filterViewPager.setAdapter(new FilterSectionAdapter(fragmentManager, fragments));

        binding.filterTabs.setupWithViewPager(binding.filterViewPager);

        TabLayout.Tab tab = binding.filterTabs.getTabAt(0);
        tab.setIcon(R.drawable.ic_schedule_black_24dp);

        tab = binding.filterTabs.getTabAt(1);
        tab.setText("r/");

        tab = binding.filterTabs.getTabAt(2);
        tab.setIcon(R.drawable.ic_person_black_24dp);

        final int colorAccent = ContextCompat.getColor(getActivity(), R.color.colorAccent);

        RxTabLayout.selectionEvents(binding.filterTabs)
                .filter(selectionEvent -> selectionEvent.tab().getIcon() != null)
                .subscribe(selectionEvent -> {
                    final TabLayout.Tab tab1 = selectionEvent.tab();
                    final Drawable icon = tab1.getIcon().mutate();
                    if (tab1.isSelected()) {
                        icon.setColorFilter(colorAccent, PorterDuff.Mode.SRC_IN);
                        icon.setAlpha(255);
                    } else {
                        icon.setColorFilter(Color.BLACK, PorterDuff.Mode.SRC_IN);
                        icon.setAlpha((int) (0.54 * 255));
                    }
                }, Timber::e);

        return binding.getRoot();
    }

    private static class FilterSectionAdapter extends FragmentStatePagerAdapter {

        private final int tabCount;
        private final Fragment[] fragments;

        FilterSectionAdapter(final FragmentManager fm, final Fragment[] fragments) {
            super(fm);
            this.fragments = fragments;

            tabCount = fragments.length;
        }

        @Override
        public Fragment getItem(final int position) {
            return fragments[position];
        }

        @Override
        public int getCount() {
            return tabCount;
        }
    }
}

File: app/src/main/java/com/veyndan/paper/reddit/FilterFragment.kt
package com.veyndan.paper.reddit

import android.content.Intent
import android.graphics.Color
import android.graphics.PorterDuff
import android.graphics.drawable.Drawable
import android.os.Bundle
import android.support.annotation.ColorInt
import android.support.design.widget.TabLayout
import android.support.v4.app.DialogFragment
import android.support.v4.app.Fragment
import android.support.v4.app.FragmentManager
import android.support.v4.app.FragmentStatePagerAdapter
import android.support.v4.content.ContextCompat
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.jakewharton.rxbinding2.support.design.widget.selectionEvents
import com.jakewharton.rxbinding2.view.clicks
import com.veyndan.paper.reddit.api.reddit.Reddit
import com.veyndan.paper.reddit.databinding.FragmentFilterBinding

class FilterFragment : DialogFragment() {

    override fun onCreateView(inflater: LayoutInflater?, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        val binding: FragmentFilterBinding = FragmentFilterBinding.inflate(inflater, container, false)

        val fragments: Array<Fragment> = arrayOf(TimePeriodFilterFragment.newInstance(), SubredditFilterFragment.newInstance(), UserFilterFragment.newInstance())

        binding.filterDone.clicks()
                .subscribe {
                    val intent: Intent = Intent(context, MainActivity::class.java)
                    for (fragment in fragments) {
                        intent.putExtra(Reddit.FILTER, (fragment as Filter).requestFilter())
                    }
                    startActivity(intent)

                    dismiss()
                }

        val fragmentManager: FragmentManager = childFragmentManager
        binding.filterViewPager.adapter = FilterSectionAdapter(fragmentManager, fragments)

        binding.filterTabs.setupWithViewPager(binding.filterViewPager)

        var tab: TabLayout.Tab = binding.filterTabs.getTabAt(0)!!
        tab.setIcon(R.drawable.ic_schedule_black_24dp)

        tab = binding.filterTabs.getTabAt(1)!!
        tab.text = "r/"

        tab = binding.filterTabs.getTabAt(2)!!
        tab.setIcon(R.drawable.ic_person_black_24dp)

        @ColorInt val colorAccent: Int = ContextCompat.getColor(activity, R.color.colorAccent)

        binding.filterTabs.selectionEvents()
                .filter { selectionEvent -> selectionEvent.tab().icon != null }
                .subscribe { selectionEvent ->
                    val tab1: TabLayout.Tab = selectionEvent.tab()
                    val icon: Drawable = tab1.icon!!.mutate()
                    if (tab1.isSelected) {
                        icon.setColorFilter(colorAccent, PorterDuff.Mode.SRC_IN)
                        icon.alpha = 255
                    } else {
                        icon.setColorFilter(Color.BLACK, PorterDuff.Mode.SRC_IN)
                        icon.alpha = (0.54 * 255).toInt()
                    }
                }

        return binding.root
    }

    private class FilterSectionAdapter internal constructor(fm: FragmentManager, private val fragments: Array<Fragment>) : FragmentStatePagerAdapter(fm) {

        private val tabCount: Int = fragments.size

        override fun getItem(position: Int): Fragment {
            return fragments[position]
        }

        override fun getCount(): Int {
            return tabCount
        }
    }

    companion object {

        fun newInstance(): FilterFragment {
            return FilterFragment()
        }
    }
}
